Added replacement string option

JohnAkerman opened the pull request:
I have added a new parameter `options` which is used to provide further customisation. In this version I have added a variable called `replaceStr` which is used to define a different replacement string. The default value is `'...'`.

I have updated the demo page to include the first button having "---" as the replacement string, we can change this if required.

The new `options` parameter will allow us to add new options, this is what I will use for the responsive true/false option later.

I have updated the Readme with an options section and tidied a few things up.

JohnAkerman committed:
Merge remote-tracking branch 'refs/remotes/nzambello/master'

JohnAkerman committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into replacement-string

JohnAkerman committed:
Added replacement string option

This includes the ability to extend the options object for further
customisation. Readme updated accordingly.

JohnAkerman committed:
Usage readme update

Removed unwanted text

pnicolli commented on src/ellipsed.js:
```diff
@@ -40,7 +40,15 @@ function tokensReducer(acc, token) {
 }
 
-function ellipsis(selector = '', rows = 1) {
+function ellipsis(selector = '', rows = 1, options) {
```
We are using 2-space indentation here, it would be nice to also have these new lines with the same indentation for consistency.

pnicolli commented on src/ellipsed.js:
```diff
@@ -40,7 +40,15 @@ function tokensReducer(acc, token) {
+    };
+
+    // Override options with any new values if defined
+    opts = Object.assign(opts, options);
```
In other parts of the code we used the object spread operator to do this, since it's clearer to read. I would suggest writing this as `opts = { ...opts, ...options };` or something like that, in order to be consistent with the rest of the code.

JohnAkerman replied:
I've actually never heard of object spread operator - I assume its new to ES6? I will update accordingly, thanks for the tip.

nzambello replied:
I'll also remove those comments and rename `opts` in `defaultOpts` or something like this, for code readability/style.

nzambello commented:
I like the options object and relative change in README. I'm leaving some comments before merge.

nzambello commented on src/ellipsed.js:
```diff
@@ -9,25 +9,25 @@
       el.innerHTML = textBeforeWrap[textBeforeWrap.length - 1] === '.'
-        ? `${textBeforeWrap}..`
-        : `${textBeforeWrap}...`;
+        ? `${textBeforeWrap}${options.replaceStr}`
```
I would leave line 29 as it is, because it was an interesting feature and now the condition is useless.
I'll change it in the merge commit.

(adding a condition on the replaceStr)

JohnAkerman replied:
Good point, the code checks to see if the last character is a `.` and if so only adds two dots, instead of the default three?

I'll revert the change.

nzambello replied:
Sorry, I made a mistake writing the line of code: what we were talking was in line 29, and if I'm not wrong, then you reverted the wrong change.

nzambello commented on example/index.js:
```diff
@@ -12,7 +12,7 @@ function reset() {
 function uno() {
   reset();
-  ellipsis('.text p', 1);
+  ellipsis('.text p', 1, { replaceStr: "---" });
```
If you add this to the example, add a label in the html, too. You have to explain to the user what's happening when he presses that button.

pnicolli reviewed (commented):
Minor comments about consistency with the rest of the library. I like the PR overall, thanks!

JohnAkerman committed:
Reverted replaceStr if last character is a dot

JohnAkerman committed:
Updated tab size to be consistent

The tab size is now 2 on my latest changes to be consistent with the
rest of the code.

JohnAkerman committed:
Added new demo controls for replaceStr option

I have added a new section on the demo to house the different options,
which can be expanded on in the future.

JohnAkerman committed:
Cleaned up object extend and reverted Line 20 change

JohnAkerman committed:
Reset test data

JohnAkerman commented:
I have updated this Pull Request with the changes you have recommended, see [above ](https://github.com/nzambello/ellipsed/pull/16#commits-pushed-67ada7f)comments. 

This includes a slightly new layout of the demo page:
![image](https://user-images.githubusercontent.com/2814624/30825073-9b7ed59c-a229-11e7-953b-12ca955a2f12.png)

nzambello commented on index.html:
```diff
@@ -30,7 +30,10 @@
                 <button class="btn" onclick="quattro()">4</button>
-                <button class="btn" onclick="cinque()">5</button>
+                <button class="btn" onclick="cinque()">5</button><br>
+                <h3>Options</h3>
```
I don't like very much this section. We could think about an input element with the replaceStr value, so the user can try with another string. But I would like to do this in an other PR because I'd like to close this and keep it clean.

What I'm doubtful about is the second reset button: they are identical and it's not clear the meaning of that option section.

We can further do this adding the <input> I was talking about and with some text that references to the replaceStr meaning. 

So, I would not change the demo in this moment.

JohnAkerman replied:
That makes sense. I have now reverted the changes to the demo page and updated the branch.

nzambello commented on example/index.js:
```diff
@@ -34,3 +34,8 @@ function cinque() {
   ellipsis('.text p', 5);
 }
+
+function unoDifferentReplacementStr() {
```
Referring to comments in `index.html`, this will be useful soon, but not in this PR.

nzambello commented on example/index.css:
```diff
@@ -34,6 +34,10 @@ h1 {
     box-sizing: border-box;
 }
 
+.controls h3 {
```
Same here as in demo template and JS.

nzambello reviewed (commented):
There are still some things to handle, but after these it is ready for merge.

JohnAkerman committed:
Reverted changes to demo

Replacement string changes to the demo page have been reverted

JohnAkerman commented:
I have now updated the branch to include the change to the character check replace and reverted the demo changes.

@nzambello

JohnAkerman committed:
Updated replaceStr

nzambello commented:
Well done, thank you @JohnAkerman.
